File: model/src/main/java/io/github/chriszhong/model/organization/event/CharacteristicEventFactory.java
package io.github.chriszhong.model.organization.event;

import io.github.runtimemodels.chazm.entity.Characteristic;

/**
 * The {@linkplain CharacteristicEventFactory} interface defines the API for constructing {@linkplain CharacteristicEvent}s.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
@FunctionalInterface
public interface CharacteristicEventFactory {

	/**
	 * Constructs an {@linkplain CharacteristicEvent}.
	 *
	 * @param category
	 *            the {@linkplain EventCategory}.
	 * @param characteristic
	 *            the {@linkplain Characteristic}.
	 *
	 * @return a {@linkplain CharacteristicEvent}.
	 */
	CharacteristicEvent build(EventCategory category, Characteristic characteristic);

}

File: chazm-model/src/main/kotlin/runtimemodels/chazm/model/entity/CharacteristicEventFactory.kt
package runtimemodels.chazm.model.entity

import runtimemodels.chazm.api.entity.Characteristic
import runtimemodels.chazm.model.event.EventType

/**
 * The [CharacteristicEventFactory] interface defines the API for constructing [CharacteristicEvent]s.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
@FunctionalInterface
interface CharacteristicEventFactory {
    /**
     * Constructs an [CharacteristicEvent].
     *
     * @param category       the [EventType].
     * @param characteristic the [Characteristic].
     * @return a [CharacteristicEvent].
     */
    fun build(category: EventType, characteristic: Characteristic): CharacteristicEvent
}
